File: src/main/java/work/nekow/particledrawing/lighting/LightAttenuation.java
package work.nekow.particledrawing.lighting;

/**
 * Defines how dynamic light brightness attenuates over distance.
 * Implementations must be thread-safe and fast (called per-light-source per-frame).
 */
@FunctionalInterface
public interface LightAttenuation {

    /**
     * Evaluates the attenuation factor at a given distance.
     *
     * @param distance    the distance from the light source, in blocks
     * @param maxDistance the maximum influence radius of this light source
     * @return attenuation factor in [0, 1], where 1 = full brightness, 0 = no contribution
     */
    float evaluate(float distance, float maxDistance);

    LightAttenuation LINEAR = (distance, maxDistance) -> {
        if (distance >= maxDistance) return 0f;
        return 1f - (distance / maxDistance);
    };

    LightAttenuation INVERSE_SQUARE = (distance, maxDistance) -> {
        if (distance >= maxDistance) return 0f;
        float t = distance / maxDistance;
        return 1f / (1f + t * t * 8f);
    };

    LightAttenuation SMOOTHSTEP = (distance, maxDistance) -> {
        if (distance >= maxDistance) return 0f;
        float t = distance / maxDistance;
        return 1f - (t * t * (3f - 2f * t));
    };

    LightAttenuation INVERSE_LINEAR = (distance, maxDistance) -> {
        if (distance >= maxDistance) return 0f;
        return 1f / (1f + distance * 2f);
    };

    /**
     * Creates a cubic-bezier-based attenuation.
     */
    static LightAttenuation bezier(double x1, double y1, double x2, double y2) {
        work.nekow.particledrawing.core.easing.EasingCurve curve =
            new work.nekow.particledrawing.core.easing.EasingCurve(x1, y1, x2, y2);
        return (distance, maxDistance) -> {
            if (distance >= maxDistance) return 0f;
            float t = distance / maxDistance;
            return 1f - curve.evaluate(t);
        };
    }
}

File: src/main/kotlin/work/nekow/particledrawing/lighting/LightAttenuation.kt
package work.nekow.particledrawing.lighting

import work.nekow.particledrawing.core.easing.EasingCurve

/**
 * 光照衰减函数接口。
 * 定义动态光源随距离衰减的强度计算方式。
 */
@Suppress("unused")
fun interface LightAttenuation {

    /**
     * 计算指定距离下的衰减因子。
     * @param distance 到光源的距离
     * @param maxDistance 光源最大有效范围
     * @return 衰减因子 (0=无光照, 1=满光照)
     */
    fun evaluate(distance: Float, maxDistance: Float): Float

    companion object {
        /** 线性衰减: 距离越远越暗 */
        @JvmField val LINEAR = LightAttenuation { distance, maxDistance ->
            if (distance >= maxDistance) 0f
            else 1f - (distance / maxDistance)
        }

        /** 平方反比衰减: 模拟真实物理光照 */
        @JvmField val INVERSE_SQUARE = LightAttenuation { distance, maxDistance ->
            if (distance >= maxDistance) 0f
            else {
                val t = distance / maxDistance
                1f / (1f + t * t * 8f)
            }
        }

        /** 平滑阶梯衰减: 使用 S 曲线过渡 */
        @JvmField val SMOOTHSTEP = LightAttenuation { distance, maxDistance ->
            if (distance >= maxDistance) 0f
            else {
                val t = distance / maxDistance
                1f - (t * t * (3f - 2f * t))
            }
        }

        /** 反比线性衰减: 近距离衰减快，远距离保持微弱光照 */
        @JvmField val INVERSE_LINEAR = LightAttenuation { distance, maxDistance ->
            if (distance >= maxDistance) 0f
            else 1f / (1f + distance * 2f)
        }

        /**
         * 使用贝塞尔曲线自定义衰减。
         * @param x1 控制点1 X
         * @param y1 控制点1 Y
         * @param x2 控制点2 X
         * @param y2 控制点2 Y
         * @return 自定义衰减函数
         */
        @JvmStatic
        fun bezier(x1: Double, y1: Double, x2: Double, y2: Double): LightAttenuation {
            val curve = EasingCurve(x1, y1, x2, y2)
            return LightAttenuation { distance, maxDistance ->
                if (distance >= maxDistance) 0f else {
                    val t = distance / maxDistance
                    1f - curve.evaluate(t)
                }
            }
        }
    }
}
